### ISSUE TITLE
ExistingSessionDetectedAction=primary not working with TWS 950.2g

### ISSUE THREAD

**Issue by berndulum** [ib-controller/ib-controller#39]
Hey,

I am using ib-controller 2.13.1 and tws 950.2g and experience problems with setting ExistingSessionDetectedAction to primary. When starting tws with ibcontroller and another session is currently active the setting works as expected and grabs the existing session. But when tws with ibcontroller is already running and I start another session the upcoming relogin dialog is not handled at all and the session is not reconnected. This is not the behavior stated in the ini files comment.

I can confirm that the setting works as expected with TWS 948.3c. Therefore I think the problem is that the reconnect dialog has changed in a recent version of tws:
![tws_reconnect](https://cloud.githubusercontent.com/assets/14736709/10013871/919799b2-6115-11e5-9f02-4f0567e9fe62.png)

**Comment by rlktradewright** [comment]
I need you to be a bit clearer about what you're doing here so I can try and reproduce the problem (so far I've failed).

Here are the possible ways of starting TWS (as far as ExistingSessionDetectedAction is concerned):

* start TWS without using IBController
* start TWS using IBController with ExistingSessionDetectedAction=primary
* start TWS using IBController with ExistingSessionDetectedAction=secondary
* start TWS using IBController with ExistingSessionDetectedAction=manual

Please detail which combination of starts leads to the problem you encountered.

Thanks.

**Comment by berndulum** [comment]
I did the following:

1. Start the first TWS (I) without IBController
2. Start a second TWS (II) with IBController with ExistingSessionDetectedAction=primary
This causes the newly started TWS (II) to grab the session from the first one.
3. Restart the first TWS (I) without IBController, a dialog appears saying there is already another TWS running. Clicking the "continue login" button causes the session to be in this TWS (I) again.
4. In the second TWS (II) the dialog shown in my first post appears. But it is not automatically handled by IBController.

TWS I and II are running on different machines, so there is no influence (like same working dir, ...) between them.

Please ask if my procedure is still unclear.

**Comment by rlktradewright** [comment]
Ok thanks for the additional info. I'll check it out as soon as I get time, but it won't be today.

You're certainly right that IBController doesn't handle the dialog you showed: I've never even seen it before!

**Comment by rlktradewright** [comment]
I've added code to handle the 'Re-login is required' dialog, and this now works fine if both TWS instances are version 950 or later.

However, there is a 'gotcha' if the instance started without IBController (or with ExistingSessionDetectedAction=manual) is pre-950. In these versions, once an instance (A) has taken over a session from another instance (B), if the other instance (B) then takes the session away again, then instance A will automatically re-logon again and take it back again. (Note, however, that if a third instance (C) takes the session away, then A will display the 'Existing session detected' dialog again). But in 950 and 952, TWS will always display the 'Existing session detected' dialog and will not re-logon automatically.

So in the scenario described above by @berndulum where instance A was TWS 948 and instance B was TWS 950, then it will be impossible for instance B to take and keep the session even with correct handling of the  'Re-login is required' dialog, because A keeps taking it back again: there is nothing I can do in IBController to prevent this, it's just the way older versions of TWS work.

However, in practice this really shouldn't be a problem for anyone. If you start a TWS instance (either manually or via IBController with ExistingSessionDetectedAction=manual), and you see the 'Existing session detected' dialog, and you know that you have an IBController set up with ExistingSessionDetectedAction=primary, then you really shouldn't try to take over the session.

Of course the best thing is to make sure you always start TWS using IBController, and that every IBController instance that is not the primary instance is explicitly set to secondary, not manual.

I'll issue a new release containing this fix later today.

**Comment by rlktradewright** [comment]
Further to my previous post, I need to add something more to describe what I've found since then.

I can only reproduce the problem described in the previous post when the following conditions apply:

1. My primary TWS (logged into my paper trading account) is using TWS 950 or greater 
2. AND the primary TWS is running in the same machine as my live TWS, so that the primary TWS can receive realtime data.
3. AND the other TWS that is trying to take over the session is running TWS 949 (or earlier)
4. AND the other TWS is not on the same machine.

If I use any other setup, then the session trying to take over always displays the 'Existing session detected' dialog, so everything is fine.

I've only added this for completeness, since it doesn't make any practical difference to the advice given in the previous post.
